tone equalizer: improve pickers

Mark-64 opened the pull request:
This PR replaces #9998 and #10014 into a clean one. 
It includes two commits:

1) Fixing erratic gui behavior of TE picker buttons (fixes #9995, fixes #9997)
corresponds to the first old PR, already reviewed by @aurelienpierre 

2) Improving the auto-exposure and auto-contrast estimations
Today they are not accurate, for two reasons:
- they are based on histogram population statistics (first and last deciles), but those are calculated on a histogram "clipped" in the range [-8, 0] EV, which is insufficient in most cases (particularly the 0 EV upper limit). The solution implemented is to use an extended histogram [-10, +6] EV for the purpose of stats calculation
- the auto-contrast estimation formula assumes that contrast correction will expand/compress the histogram simmetrically around the fulcrum in the EV space. This is true if the contrast is a log one, however TE uses a linear contrast and the former does not hold.

Therefore I implemented a new estimation math, as follows:
In the linear space (so all variable listed below are `exp2f()` of their EV counterparts) the exposure/contrast correction are applied by TE in this way

![immagine](https://user-images.githubusercontent.com/43290988/133918408-84e894d6-8da0-4ded-ad83-9b6d8cfab8d8.png)

where `l` and `l'` are the luminance before and after correction, `e` is the exposure and `c` is the contrast

We want

![immagine](https://user-images.githubusercontent.com/43290988/133918456-d908641a-a911-48d6-8ab1-10edd5f4e47f.png)

where `l1`, `l2` are the first and last deciles, `t1`, `t2` are the target values for them (-7 EV and -1 EV). Strictly speaking, the above formulas are valid for single points, not necessarily for population stats like deciles, but the approximation is pretty good

Solving for c and e we get:

![immagine](https://user-images.githubusercontent.com/43290988/133918534-35dae798-cff9-4021-942d-fd51e6d449b5.png)

![immagine](https://user-images.githubusercontent.com/43290988/133918546-221a3370-ec76-46da-9b2d-5e811ee80c7c.png)

By the way I replaced deciles with "ventiles" (histogram population <5% or >95%), because the auto correction looks more accurate

TurboGit commented on src/iop/toneequal.c:
```diff
@@ -1725,36 +1725,33 @@ static void smoothing_callback(GtkWidget *slider, gpointer user_data)
+  if(!g->luminance_valid || self->dev->pipe->processing || !g->histogram_valid)
   {
+    ++darktable.gui->reset;
+    gboolean active = dt_bauhaus_widget_get_quad_active(quad);
```
const

Mark-64 replied:
Ok, I will replace the pickers with a regular non-toggle button, possibly with a new icon. Maybe a simple "A" standing for "auto", or something resembling that.

TurboGit commented on src/iop/toneequal.c:
```diff
@@ -1348,67 +1349,67 @@ static inline void build_interpolation_matrix(float A[CHANNELS * PIXEL_CHAN],
+  for(size_t k = 0; k < TEMP_SAMPLES; ++k)
+  {
+    float EV = 16.0 * (float)k / (float)(TEMP_SAMPLES - 1) - 10.0;
+    int i = CLAMP((int)(((EV + 8.0f) / 8.0f) * (float)UI_SAMPLES), 0, UI_SAMPLES - 1);
```
const

TurboGit commented on src/iop/toneequal.c:
```diff
@@ -1348,67 +1349,67 @@ static inline void build_interpolation_matrix(float A[CHANNELS * PIXEL_CHAN],
-  for(size_t k = 0; k < hist_bins; ++k)
+  // scout the extended histogram bins looking for deciles
+  // these would not be accurate with the regular histogram
+  for(size_t k = 0; k < TEMP_SAMPLES; ++k)
```
Would an omp for directive will speed-up things here?

Mark-64 replied:
I don't think so, for simple math on an array of 256 integers (512 for the temp histogram). 
Moreover, I am not sure this cycle with an accumulator can be vectorized at all. I can slightly optimize by using two cycles, one fron the start of the array, looking for first decile and another from the end, looking for last decile, stopping the cycle when they are found. But it is very minor.

TurboGit commented on src/iop/toneequal.c:
```diff
@@ -1348,67 +1349,67 @@ static inline void build_interpolation_matrix(float A[CHANNELS * PIXEL_CHAN],
+  *max_histogram = 0;
+  // remap the extended histogram into the normal one
+  // bins between [-8; 0] EV remapped between [0 ; UI_SAMPLES]
+  for(size_t k = 0; k < TEMP_SAMPLES; ++k)
```
Same as above. Would an omp for directive will speed-up things here?

Mark-64 replied:
same as above. I can make a test, just for curiosity

aurelienpierre replied:
OpenMP can't be used here, it would create a race condition. We could use `#pragma omp for reduction(+:histogram)` to avoid it if `histogram` was a scalar, but here it's an array and OpenMP didn't implement reductions on arrays/vector last time I checked.

TurboGit commented on src/iop/toneequal.c:
```diff
@@ -1787,53 +1798,59 @@ static void auto_adjust_exposure_boost(GtkWidget *quad, gpointer user_data)
+  if(!g->luminance_valid || self->dev->pipe->processing || !g->histogram_valid)
   {
+    ++darktable.gui->reset;
+    gboolean active = dt_bauhaus_widget_get_quad_active(quad);
```
const

TurboGit reviewed (changes requested):
Some minor changes and a possible perf opt.

Another point to raise as we are on it. The color picker in the tone equalizer are not real ones. You cannot select an area, it is kind of "auto"  tune based on the picture. What annoy me is that the color picker are left selected after clicking and doing their job. This is not good. I would at least deselect them when the adjustment is done. And we could even maybe come up with a new icon for those auto picker which do not allow for selecting an area?

Mark-64 commented:
with the last commit:
- changed the buttons icon from eyedropper to "auto levels" from retouch module, for now. If this is merged we can think of a better icon.
- the button are no longer toggle 
- minor optimization of the histogram loops. I tried to parallelize, but it is slower, for such small arrays
- added const as requested

The behavior of the button is still not fantastic because bauhaus buttons next to sliders are actually made for being toggle. When set as no toggle they do not provide any feedback on click to the user, and no prelight either. They almost seem inactive.
I don't know if there is a way to place a normal button in the same place.
Moreover, with the previous implementation, the toggle button had the subtle function to reset the correction and recalculate the histogram and the relative deciles, which means that a new auto-exposure/contrast calculation will always restart from the same data.
Without the toggle, the previous correction must be reverted "in theory" by the auto-correction math, which is what is done, but it's not perfect.
This is because the effect of the surface blur on the histogram is a bit unpredictable, especially for eigf (see also #10034) so it cannot be reverted accurately without actually recalculate the histogram.
The bottom line effect is that clicking mutliple time on the buttons lead to slighlty different estimations, especially when a contrast is dialed in.

Mark-64 commented:
small modification in bauhaus.c to highlight the quad button when clicked in non-toggle mode, in order to provide visual feedback to the user

TurboGit commented:
Ok, to me this is good to go. I'm not fond of the icon at all but this can be worked later when we'll get more feedback. I'm not in a position to propose a better icon for now anyway.
